Add system state backup to talpid core.

pinkisemils opened the pull request:
I've added the code that will be used by the _Firewall_ trait implementors on Linux and Windows and some tests. Whilst it will only be used to store the DNS servers used by the system when a tunnel is not connected, I chose to not call it _DNSServerListBackup_ as this code really doesn't have to care about what it stores.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)
<!-- Reviewable:end -->

faern commented:
Review status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @jvff, @faern, and @pinkisemils)

---

*[talpid-core/Cargo.toml, line 38 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWKB-DWTN-o_hlHxF:-LEiWKB09FGAeK2U7ioA:b-xmh262) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/Cargo.toml#L38)):*
> ```TOML
> 
> [dev-dependencies]
> tempfile = "3.0"
> ```

Would it be possible to use `tempdir` instead? We already have that in other parts of the product, so it would not pull in another dependency for doing virtually the same thing. Or is this specific use case not covered by `tempdir`?

---

*[talpid-core/src/firewall/system_state.rs, line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWqc36ilAQqlzXzpJ:-LEiWqc36ilAQqlzXzpK:b-is276w) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L12)):*
> ```Rust
> 
> /// A callback for writing system state data
> pub extern "system" fn write_system_state_backup_cb(
> ```

It feels strange to expose this as an FFI function at this level. We will need this for Linux as well, but we don't need it to be FFI. Could we not expose this as pure Rust at the generic level. And then write an FFI version that calls the pure Rust implementation inside the windows module, which is the only one who will need an FFI version of it anyway?

---

*[talpid-core/src/firewall/system_state.rs, line 44 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXm6K1z59Hm2hVjQ9:-LEiXm6K1z59Hm2hVjQA:b-auluje) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L44)):*
> ```Rust
> 
> /// This struct is responsible for saving a binary blob to disk. The binary blob is intended to
> /// store system state that should be resotred when we reset our security policy.
> ```

s/resotred/restored/

Maybe change "when we reset our..." to "when the security policy is being reset"? Avoiding "we", "I", "our" etc is usually good in documentation.

---

*[talpid-core/src/firewall/system_state.rs, line 47 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXWwZ5ItJfkzyF0VD:-LEiXWwZ5ItJfkzyF0VE:bdqe4l4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L47)):*
> ```Rust
> pub struct SystemStateWriter {
>     /// Full path to the system state backup file
>     pub backup_path: Box<Path>,
> ```

Don't we want a `PathBuf`? That is the type for owned paths. Same relationship as `String` has to `str`.

---

*[talpid-core/src/firewall/system_state.rs, line 50 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXHUQ51wDi5hFvb0M:-LEiXHUQ51wDi5hFvb0N:bmdfeph) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L50)):*
> ```Rust
> }
> 
> const STATE_BACKUP_FILENAME: &str = "system_state_backup";
> ```

I personally prefer putting all `const`s at the top of the file, just under imports. To show the parameters controlling the execution in that module early on.

---

*[talpid-core/src/firewall/system_state.rs, line 69 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiYVWvD8J3Skit3jbe:-LEiYVWvD8J3Skit3jbf:b3586lm) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L69)):*
> ```Rust
>     }
> 
>     /// Checks to see if the previous
> ```

This documentation seems unfinished?

---

*[talpid-core/src/firewall/system_state.rs, line 74 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiYkDQ3Qe2twmtqfF9:-LEiYkDQ3Qe2twmtqfFA:biubxe5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L74)):*
> ```Rust
>         // Creating the file in a separate scope to ensure it's closed before we try to remove it,
>         // otherwise it will fail to be removed on some platforms.
>         match File::open(&self.backup_path) {
> ```

Might be able to use the `fs:.read` (https://doc.rust-lang.org/stable/std/fs/fn.read.html) function instead. Convenient shorthand for `File::open` and `read_to_end`. I also think it does use `BufReader` internally(?), which would be nice.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

jvff commented:
Reviewed 4 of 4 files at r1.
Review status: 0 of 4 files reviewed, 9 unresolved discussions (waiting on @faern, @jvff, and @pinkisemils)

---

*[talpid-core/src/firewall/system_state.rs, line 7 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXpN-9VuQ_z_NMms5:-LEiXpN1C8rE6e8o-FNF:bx7634e) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L7)):*
> ```Rust
> use std::fs::File;
> use std::io;
> use std::io::{Read, Write};
> ```

Minor (subjective) nitpick, but I think using `std::fs::{self, File}` and `std::io::{self, Read, Write}` is a little shorter. Just a comment, feel free to leave it as you prefer.

---

*[talpid-core/src/firewall/system_state.rs, line 33 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiZK5k1NrskKeGFjTQ:-LEiZK5k1NrskKeGFjTR:b-zbutsu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L33)):*
> ```Rust
>             Err(e) => {
>                 error!(
>                     "Failed write system state backup to {} because {}",
> ```

Missing a `to` before `write`?

---

*[talpid-core/src/firewall/system_state.rs, line 53 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiZXuB-kSpWGpq-v_h:-LEiZXuB-kSpWGpq-v_i:b-a2nvn2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L53)):*
> ```Rust
> 
> impl SystemStateWriter {
>     /// Creates a new SystemSateWriter which will use a file in the cache directory to store system
> ```

Typo: `SystemStateWriter`.

---

*[talpid-core/src/firewall/system_state.rs, line 183 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEicKvu6pwSK3QIBL9a:-LEicKvu6pwSK3QIBL9b:bq0t15r) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L183)):*
> ```Rust
>         writer
>             .remove_state_file()
>             .expect("Failed to remove state file when one did not exist");
> ```

Suggestion: also assert that reading the state afterwards results in `None`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Review status: 0 of 4 files reviewed, 9 unresolved discussions (waiting on @faern, @jvff, and @pinkisemils)

---

*[talpid-core/Cargo.toml, line 38 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWKB-DWTN-o_hlHxF:-LEi_c0gANB-WoIhPlb3:bu1yxnf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/Cargo.toml#L38)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Would it be possible to use `tempdir` instead? We already have that in other parts of the product, so it would not pull in another dependency for doing virtually the same thing. Or is this specific use case not covered by `tempdir`?
</blockquote></details>

`tempdir` is deprecated and replaced by `tempfile`. I'd rather we moved to `tempfile`.

---

*[talpid-core/src/firewall/system_state.rs, line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWqc36ilAQqlzXzpJ:-LEi_hfxDKB2RtNyUU3O:b-um5x3g) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L12)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It feels strange to expose this as an FFI function at this level. We will need this for Linux as well, but we don't need it to be FFI. Could we not expose this as pure Rust at the generic level. And then write an FFI version that calls the pure Rust implementation inside the windows module, which is the only one who will need an FFI version of it anyway?
</blockquote></details>

This callback is intended to only be used across FFI boundaries. On Linux, there is no reason to use this callback. `SystemStateWriter` can be passed around directly or the consumer of this module can create an ad-hoc closure to be used as a callback if required.

---

*[talpid-core/src/firewall/system_state.rs, line 33 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiZK5k1NrskKeGFjTQ:-LEie7gD7mba2TTyI2MZ:bb77uci) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L33)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Missing a `to` before `write`?
</blockquote></details>

Yep!

---

*[talpid-core/src/firewall/system_state.rs, line 44 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXm6K1z59Hm2hVjQ9:-LEie94N9_qj7XCiNWFh:bo6apac) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L44)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

s/resotred/restored/

Maybe change "when we reset our..." to "when the security policy is being reset"? Avoiding "we", "I", "our" etc is usually good in documentation.
</blockquote></details>

Indeed.

---

*[talpid-core/src/firewall/system_state.rs, line 47 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXWwZ5ItJfkzyF0VD:-LEi_sqE6zksYhvfxDzD:bq87a4d) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L47)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Don't we want a `PathBuf`? That is the type for owned paths. Same relationship as `String` has to `str`.
</blockquote></details>

The intention here is to make the path immutable.

---

*[talpid-core/src/firewall/system_state.rs, line 53 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiZXuB-kSpWGpq-v_h:-LEieNNDDAdHYtXwHILz:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L53)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Typo: `SystemStateWriter`.
</blockquote></details>

Done.

---

*[talpid-core/src/firewall/system_state.rs, line 69 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiYVWvD8J3Skit3jbe:-LEieQNB8g3tdNtqfVWc:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L69)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This documentation seems unfinished?
</blockquote></details>

Done.

---

*[talpid-core/src/firewall/system_state.rs, line 74 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiYkDQ3Qe2twmtqfF9:-LEidBUZDHWhxZv8M1mL:b-xvq7o9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L74)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Might be able to use the `fs:.read` (https://doc.rust-lang.org/stable/std/fs/fn.read.html) function instead. Convenient shorthand for `File::open` and `read_to_end`. I also think it does use `BufReader` internally(?), which would be nice.
</blockquote></details>

Nice, I was not aware of this. I also replaced `File::create()` and `File::write_all` with `fs::write`.

---

*[talpid-core/src/firewall/system_state.rs, line 183 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEicKvu6pwSK3QIBL9a:-LEidJe-9Em-Z3QM3Diy:b-896fja) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L183)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Suggestion: also assert that reading the state afterwards results in `None`.
</blockquote></details>

Done!

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @faern, @jvff, and @pinkisemils)

---

*[talpid-core/Cargo.toml, line 38 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWKB-DWTN-o_hlHxF:-LEihCIL1OgGyMZqybdk:b-gsu5va) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/Cargo.toml#L38)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

`tempdir` is deprecated and replaced by `tempfile`. I'd rather we moved to `tempfile`.
</blockquote></details>

Good find! Yes, let's move over everything to `tempfile` when given time then.

---

*[talpid-core/src/firewall/system_state.rs, line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWqc36ilAQqlzXzpJ:-LEihQg2Cze8AddTFGF0:b81fnwp) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L12)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

This callback is intended to only be used across FFI boundaries. On Linux, there is no reason to use this callback. `SystemStateWriter` can be passed around directly or the consumer of this module can create an ad-hoc closure to be used as a callback if required.
</blockquote></details>

Ok. But still, the FFI conversion is specific to one implementation, and not the module at large. So the link FFI->Rust still belongs in the Windows module IMO. And maybe the docs should say that it's a helper for FFI implementer? At the moment I get the impression that this is **the way** to write system state backups.

---

*[talpid-core/src/firewall/system_state.rs, line 47 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXWwZ5ItJfkzyF0VD:-LEii52_-gK6fuSrgDmM:bphke2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L47)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

The intention here is to make the path immutable.
</blockquote></details>

Ok. Yeah I guess the result is basically the same, but enforced immutability. I have not thought about it that way. I think we have pretty many places where we store `PathBuf` in structs and that are not supposed to be mutated.

---

*[talpid-core/src/firewall/system_state.rs, line 74 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiYkDQ3Qe2twmtqfF9:-LEiihAA-mFVgZoYWBf2:bccpe8g) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L74)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Nice, I was not aware of this. I also replaced `File::create()` and `File::write_all` with `fs::write`.
</blockquote></details>

Awesome.

---

*[talpid-core/src/firewall/system_state.rs, line 67 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiiStz9BOzjB1anehu:-LEiiStz9BOzjB1anehv:bgg9fvf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L67)):*
> ```Rust
>     }
> 
>     /// Tries to read a previously saved backup and delete's it after reading it if it exists.
> ```

"deletes"? It's not an ownership *s* nor an abbreviation for "delete is" :P

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @faern, @jvff, and @pinkisemils)

---

*[talpid-core/src/firewall/system_state.rs, line 176 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiwPDb7Lw8xOTOSEXN:-LEiwPDc3dHeGZ05DQp1:b62rhb3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L176)):*
> ```Rust
>         writer
>             .remove_state_file()
>             .expect("Failed to remove state file when one did not exist");
> ```

What does it mean "when one did not exist"? Does a backup not exist here? You wrote one just one line before?

---

*[talpid-core/src/firewall/system_state.rs, line 182 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiwa91ByjzlJaoWhJb:-LEiwa925Pj7jHwiJFiW:b-axh5k1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L182)):*
> ```Rust
>     #[test]
>     fn cant_remove_backup_with_io_error() {
>         let temp_dir = PathBuf::from("/dev/null");
> ```

This is an **existing** file, and not a non-existing dir :P Does this test what we want it to test? Maybe a non-existing dir will just give `NotFound` and be ignored I guess.

Would it be possible to make it a cross-platform test in some way?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
Review status: 0 of 4 files reviewed, 6 unresolved discussions (waiting on @jvff, @faern, and @pinkisemils)

---

*[talpid-core/src/firewall/system_state.rs, line 12 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiWqc36ilAQqlzXzpJ:-LEjKXS896TjKu6ZdyYb:bl1gk8y) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L12)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Ok. But still, the FFI conversion is specific to one implementation, and not the module at large. So the link FFI->Rust still belongs in the Windows module IMO. And maybe the docs should say that it's a helper for FFI implementer? At the moment I get the impression that this is **the way** to write system state backups.
</blockquote></details>

For now, I'm deleting the function, but I do genuinely believe that it should reside in this file, as it is specific to `SystemStateWriter`. 

---

*[talpid-core/src/firewall/system_state.rs, line 67 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiiStz9BOzjB1anehu:-LEjOflQCLTRbSv6WdDr:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L67)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

"deletes"? It's not an ownership *s* nor an abbreviation for "delete is" :P
</blockquote></details>

Done.

---

*[talpid-core/src/firewall/system_state.rs, line 176 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiwPDb7Lw8xOTOSEXN:-LEjOuk176Yq_Snf02T_:bn0tve1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L176)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

What does it mean "when one did not exist"? Does a backup not exist here? You wrote one just one line before?
</blockquote></details>

Indeed, this was not the best error message for this case.

---

*[talpid-core/src/firewall/system_state.rs, line 182 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiwa91ByjzlJaoWhJb:-LEj2FF61wlvjIOoZ_XY:b-5xwkpc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L182)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This is an **existing** file, and not a non-existing dir :P Does this test what we want it to test? Maybe a non-existing dir will just give `NotFound` and be ignored I guess.

Would it be possible to make it a cross-platform test in some way?
</blockquote></details>

What I'm really testing for is that errors besides `io::ErrorKind::NotFound` get through whilst `NotFound` is dealt with appropriately. If you have a good suggestion as to what directory would always fail to read or write on Windows, I'd be glad to make this test platform-independent.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @jvff and @faern)

---

*[talpid-core/src/firewall/system_state.rs, line 182 at r2](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiwa91ByjzlJaoWhJb:-LEjkVyDEhpdUJG8FZrf:bs7fhwx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/071715e7a531a70e8714479b37c04555f72b7bf4/talpid-core/src/firewall/system_state.rs#L182)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

What I'm really testing for is that errors besides `io::ErrorKind::NotFound` get through whilst `NotFound` is dealt with appropriately. If you have a good suggestion as to what directory would always fail to read or write on Windows, I'd be glad to make this test platform-independent.
</blockquote></details>

Yeah true. I don't know a better one really. Except maybe creating your own and locking down the permissions. But that might be complicated. At least more complicated than it's worth for this type of test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

jvff commented:
Reviewed 2 of 2 files at r3.
Review status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @faern and @jvff)

---

*[talpid-core/src/firewall/system_state.rs, line 47 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXWwZ5ItJfkzyF0VD:-LElGsX05qFzs-69izQf:b2nhoia) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L47)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Ok. Yeah I guess the result is basically the same, but enforced immutability. I have not thought about it that way. I think we have pretty many places where we store `PathBuf` in structs and that are not supposed to be mutated.
</blockquote></details>

TIL

---

*[talpid-core/src/firewall/system_state.rs, line 183 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEicKvu6pwSK3QIBL9a:-LElHPrx1Ysees056zQy:bkllj39) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L183)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Done!
</blockquote></details>

You added the assertion to `can_remove_when_no_backup_exists` (which is a good idea as well), but I was originally thinking that it should be (also) added to `can_remove_backup`. Does it make sense, or is it redundant?

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
This module is not used by anyone yet, which is why I have questions about how it will be used.

I guess the plan is that `Firewall::new()` will try to consume any existing backup and restore it. Thus automatically restoring on `Firewall` initialization?

---

Reviewed 2 of 4 files at r1, 1 of 2 files at r3.
Review status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @faern, @jvff, and @pinkisemils)

---

*[talpid-core/src/firewall/system_state.rs, line 3 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEmdOJL4L19p8sQtGZE:-LEmdOJL4L19p8sQtGZF:b-4a0z3k) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L3)):*
> ```Rust
> /// A writer for a blob that would persistently store the system state. Useful
> /// for when the application of a secuirty policy proves to be persistent across
> /// reboots
> ```

Is this doc comment attached to `use std::fs;`? If it's intended to be a module comment I think it should start with `//!`?

---

*[talpid-core/src/firewall/system_state.rs, line 30 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEme3wW6AA37_03QRhN:-LEme3wX9uGot3AbPGOF:b-x0ki17) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L30)):*
> ```Rust
> 
>     /// Writes a binary blob representing the system state before any security policies were
>     /// applied to a specified backup location.
> ```

"... any security policies were applied to a specified backup location" sounds like the policies are applied to a backup location. Maybe reword as "Writes the binary backup data to disk before any security policies are applied"?

Also removed "specified" since the backup location is not really specified, at least not to this method :)

---

*[talpid-core/src/firewall/system_state.rs, line 36 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEmeXPvDMe4fWF6VRIR:-LEmeXPvDMe4fWF6VRIS:b-lx8nj9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L36)):*
> ```Rust
> 
>     /// Tries to read a previously saved backup and deletes it after reading it if it exists.
>     pub fn consume_state_backup(&self) -> io::Result<Option<Vec<u8>>> {
> ```

So here we remove the backup upon reading it, before we have had any chance to try to restore it. Is the plan that if `consume_state_backup` succeeds but applying/restoring whatever was in there fails then we lose the backup? Since the file has already been deleted at that point.

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-:-LEmezIG5GKDHd-beyq8:bsvw8ra)*
<!-- Sent from Reviewable.io -->

pinkisemils commented:
In my case it's a member of the `WindowsFirewall` struct. This is because the firewall struct on Windows will have to pass the callback and the context pointer for the FFI callback.

---

Review status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @jvff, @faern, and @pinkisemils)

---

*[talpid-core/src/firewall/system_state.rs, line 47 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEiXWwZ5ItJfkzyF0VD:-LEn6lVj9iSZLIGQeJNI:b-bvha1) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L47)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

TIL
</blockquote></details>

The same is applicable for `str`.

---

*[talpid-core/src/firewall/system_state.rs, line 183 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEicKvu6pwSK3QIBL9a:-LEn4E5IA0rovuvsU1_d:b-rx3v1v) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L183)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

You added the assertion to `can_remove_when_no_backup_exists` (which is a good idea as well), but I was originally thinking that it should be (also) added to `can_remove_backup`. Does it make sense, or is it redundant?
</blockquote></details>

I wish we had some sort of a code coverage tool so that we could make informed decisions about what else should be tested. (_not saying that coverage means that tests make sense_) But yes, I think it would make sense to do that as well.

---

*[talpid-core/src/firewall/system_state.rs, line 3 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEmdOJL4L19p8sQtGZE:-LEnAsp95IhSaw5O1UJK:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L3)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Is this doc comment attached to `use std::fs;`? If it's intended to be a module comment I think it should start with `//!`?
</blockquote></details>

Done.

---

*[talpid-core/src/firewall/system_state.rs, line 30 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEme3wW6AA37_03QRhN:-LEn7QaOAwLd6LRNOF3p:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L30)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

"... any security policies were applied to a specified backup location" sounds like the policies are applied to a backup location. Maybe reword as "Writes the binary backup data to disk before any security policies are applied"?

Also removed "specified" since the backup location is not really specified, at least not to this method :)
</blockquote></details>

Done.

---

*[talpid-core/src/firewall/system_state.rs, line 36 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEmeXPvDMe4fWF6VRIR:-LEn5uIz1f6Vxib0hGoE:bbzsqww) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/3b947769f64d37463c23ff07477ef8278a3a69a6/talpid-core/src/firewall/system_state.rs#L36)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

So here we remove the backup upon reading it, before we have had any chance to try to restore it. Is the plan that if `consume_state_backup` succeeds but applying/restoring whatever was in there fails then we lose the backup? Since the file has already been deleted at that point.
</blockquote></details>

Yes. If we can't restore a backup once, will we be successful if we try again? In any way, if we want to keep hold of the backup, as long as there is no panic, we can still write the backup back to disk in the case of an error, but that is of course up to the consumer of this struct. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-:-LEn3Z6nAZkJPgJIOz97:b-pls00k)*
<!-- Sent from Reviewable.io -->

jvff commented:
Reviewed 1 of 1 files at r4.
Review status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @faern and @jvff)

---

*[talpid-core/src/firewall/system_state.rs, line 183 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204#-LEicKvu6pwSK3QIBL9a:-LEnzMHLBLgRCnZAdT6F:b-djy053) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/75bedb0f0c8ac00db349f265fb8ed89d2a3d9603/talpid-core/src/firewall/system_state.rs#L183)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

I wish we had some sort of a code coverage tool so that we could make informed decisions about what else should be tested. (_not saying that coverage means that tests make sense_) But yes, I think it would make sense to do that as well.
</blockquote></details>

Yes, I guess in this case the code coverage tool could've been useful. Anyway, better to cover it twice just in case :smile:  

---


*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 1 files at r4.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 2 of 2 files at r5.
Review status: 2 of 4 files reviewed, all discussions resolved (waiting on @jvff and @faern)

---



*Comments from [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/204)*
<!-- Sent from Reviewable.io -->
